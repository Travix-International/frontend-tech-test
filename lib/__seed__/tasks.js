export const tasks = [
  {
    _id: '1',
    title: 'Wash the dishes',
    description: 'Yet another generic task',
    isComplete: false
  },
  {
    _id: '2',
    title: 'Do the laundry',
    description: 'Existence is pain',
    isComplete: true
  },
  {
    _id: '3',
    title: 'Take over the world',
    description: 'Send an invite to Gru and Megamind',
    isComplete: false
  }
];
