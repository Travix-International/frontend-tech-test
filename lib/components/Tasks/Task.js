import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

import styles from './Task.scss';

class Task extends PureComponent {
  static propTypes = {
    _id: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    isComplete: PropTypes.bool.isRequired,
    onEdit: PropTypes.func.isRequired,
    onRemove: PropTypes.func.isRequired,
    onToggle: PropTypes.func.isRequired,
    title: PropTypes.string.isRequired
  };

  toggleTask = () => {
    const { onToggle, _id, isComplete } = this.props;
    onToggle(_id, !isComplete);
  };

  removeTask = () => {
    const { onRemove, _id } = this.props;
    onRemove(_id);
  };

  editTask = () => {
    const { onEdit, _id } = this.props;
    onEdit(_id);
  };

  render() {
    const { title, description, isComplete, _id } = this.props;
    const taskStyles = `${styles.root} ${isComplete ? styles.complete : ''}`;
    const checkboxStyles = `${styles.checkbox} ${
      isComplete ? styles.checked : ''
    }`;
    return (
      <li className={taskStyles}>
        <div className={checkboxStyles}>
          <label htmlFor={`todo-${_id}`}>{title}</label>
          <input
            checked={isComplete}
            id={`todo-${_id}`}
            name="checkbox"
            onChange={this.toggleTask}
            type="checkbox"
          />
        </div>
        {description.length !== 0 && (
          <p className={styles.description}>{description}</p>
        )}
        <div className={styles.options}>
          <button className={styles.edit} onClick={this.editTask}>
            Edit
          </button>
          <button className={styles.delete} onClick={this.removeTask}>
            Delete
          </button>
        </div>
      </li>
    );
  }
}

export default Task;
