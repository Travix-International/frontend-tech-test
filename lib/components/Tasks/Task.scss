@import '../../scss/_settings.scss';
@import '../../scss/_objects.scss';

.root {
  background: #dc1c27;
  background: linear-gradient(
    to bottom,
    $color-brand 0%,
    $color-brand-light 100%
  );
  box-shadow: 0 10px 20px rgba(0, 0, 0, 0.19), 0 6px 6px rgba(0, 0, 0, 0.23);
  transition: all 0.3s cubic-bezier(0.25, 0.8, 0.25, 1);
  border-radius: 4px;
  color: $color-light;
  padding: 20px 15px;
  position: relative;
  margin-bottom: $spacing-1;

  &:hover {
    box-shadow: 0 14px 28px rgba(0, 0, 0, 0.25), 0 10px 10px rgba(0, 0, 0, 0.22);
  }
}

.checkbox {
  @extend .checkbox;

  label {
    font-size: 1.5em;

    &::before {
      top: 9px;
    }

    &::after {
      top: 14px;
    }
  }
}

.checked {
  @extend .checkbox--checked;
}

.complete {
  background: $color-secondary;
}

.delete {
  @extend .text-button;
  bottom: 20px;
  color: $color-light;
  font-size: 13px;
  font-weight: 300;
  position: absolute;
  right: 20px;
}

.description {
  max-width: 85%;
}
