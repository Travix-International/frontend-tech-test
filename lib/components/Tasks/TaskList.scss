@import '../../scss/_settings.scss';

.root {
  margin: $spacing-1 0;
}

.list {
}

.message {
}
