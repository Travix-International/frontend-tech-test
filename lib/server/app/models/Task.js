import mongoose from 'mongoose';

const Schema = mongoose.Schema;

const TaskSchema = new Schema({
  title: { type: String, required: true, minlength: 1, trim: true },
  description: { type: String, trim: true },
  isComplete: { type: Boolean, default: false }
});

const Task = mongoose.model('Task', TaskSchema);

export default Task;
