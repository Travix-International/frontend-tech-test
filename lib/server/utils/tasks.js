import axios from 'axios';

/* Helper to get tasks using GET API */
export const fetchTasks = async () => {
  const requestURL = `${process.env.HOST}:${process.env.PORT}`;
  const response = await axios.get(`${requestURL}/api/tasks`);
  return response.data;
};
