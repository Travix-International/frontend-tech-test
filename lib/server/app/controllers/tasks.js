import pick from 'lodash.pick';
import emitter from 'central-event';

import Task from '../models/Task';

/**
 * Helper middleware for
 * PUT/PATCH/DELETE /api/tasks/:id
 *
 * id: ObjectId
 *
 * Gets the requested task by its ID from db and sets it on the request object.
 * Thus avoiding repeating the db request logic.
 *
 * If found return status code 200 and the resource.
 * If not found return status code 404.
 */
export const helperMiddleware = async (req, res, next) => {
  const task = await Task.findById(req.params.id);
  if (task) {
    req.task = task;
    next();
  } else {
    res.status(404).send({ message: 'Not found' });
  }
};

/**
 * GET /api/tasks
 *
 * Return the list of tasks with status code 200.
 */
export const fetch = async (req, res) => {
  try {
    const tasks = await Task.find({});
    res.status(200).json(tasks);
  } catch (error) {
    res.status(500).send({
      message: error.message
    });
  }
};

/**
 * POST /api/tasks
 *
 * title: string
 * description: string
 *
 * Add a new task to the DB with the given title and description.
 * Notify Websockets server to push new data to clients
 * Return status code 201.
 */
export const create = (req, res) => {
  const task = new Task(req.body);
  // TODO: Sanitize title, description
  if (!req.body.title) {
    res.status(400).send({ message: 'Task is required' });
  } else {
    task.save();
    emitter.emit('tasks/updated');
    res.status(201).send(task);
  }
};

/**
 * PATCH /api/tasks/:id
 *
 * id: ObjectId
 * title: string
 * description: string
 * isComplete: boolean
 *
 * Update the task with the given id.
 * Notify Websockets server to push new data to clients
 * If the task is found and updated as well, return a status code 200.
 */
export const update = async (req, res) => {
  if (req.body._id) {
    delete req.body._id;
  }

  Object.keys(req.body).forEach((prop) => {
    req.task[prop] = req.body[prop];
  });

  try {
    await req.task.save();
    emitter.emit('tasks/updated');
    res.status(200).json(req.task);
  } catch (error) {
    res.status(500).send({ message: error.message });
  }
};

/**
 * PUT /api/tasks/:id
 *
 * id: ObjectId
 * title: string
 * description: string
 *
 * Update the task with the given id.
 * Notify Websockets server to push new data to clients
 * If the task is found and updated as well, return a status code 200.
 */
export const replace = async (req, res) => {
  Object.assign(
    req.task,
    pick(req.body, ['text', 'description', 'isComplete'])
  );

  try {
    await req.task.save();
    emitter.emit('tasks/updated');
    res.status(200).json(req.task);
  } catch (error) {
    res.status(500).send({ message: error.message });
  }
};

/**
 * DELETE /task/:id
 *
 * id: ObjectId
 *
 * Delete the task linked to the  given id.
 * Notify Websockets server to push new data to clients
 * If the task is found and deleted as well, return a status code 204.
 */
export const remove = async (req, res) => {
  try {
    await req.task.remove();
    emitter.emit('tasks/updated');
    res.status(204).send({ message: 'Removed' });
  } catch (error) {
    res.status(500).send({ message: error.message });
  }
};
