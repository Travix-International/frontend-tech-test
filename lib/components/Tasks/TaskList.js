import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux'; 

import styles from './TaskList.scss';
import Task from './Task';
import * as taskActions from '../../actions/taskActions';
import { sortTasks, getVisibleTasks } from './utils';

export class TaskList extends Component {
  static propTypes = {
    actions: PropTypes.object.isRequired,
    showCompleted: PropTypes.bool.isRequired,
    tasks: PropTypes.array.isRequired
  };

  toggleTask = (_id, isComplete) => {
    this.props.actions.updateTask(_id, {
      isComplete
    });
  };

  removeTask = (_id) => {
    this.props.actions.removeTask(_id);
  };

  editTask = (_id) => {
    this.props.actions.toggleEditModal(_id);
  };

  render() {
    const { tasks, showCompleted } = this.props;
    /* always show incomplete tasks first */
    const visibleTasks = sortTasks(
      showCompleted ? tasks : getVisibleTasks(tasks)
    );
    if (visibleTasks.length === 0) {
      const content = <p className={styles.message}>Nothing to do</p>;
      return <div className={styles.root}>{content}</div>;
    }

    /* eslint-disable no-underscore-dangle */
    return (
      <div className={styles.root}>
        <ul className={styles.list}>
          {visibleTasks.map(task => (
            <Task
              key={task._id} 
              {...task}
              onEdit={this.editTask}
              onRemove={this.removeTask}
              onToggle={this.toggleTask}
            />
          ))}
        </ul>
      </div>
    );
  }
}

const mapStateToProps = state => state;

const mapDispatchToProps = dispatch => ({
  actions: bindActionCreators(taskActions, dispatch)
});

export default connect(mapStateToProps, mapDispatchToProps)(TaskList);
