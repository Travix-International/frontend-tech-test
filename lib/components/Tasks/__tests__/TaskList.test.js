import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import * as tasksSeed from '../../../__seed__/tasks';
import { TaskList } from '../TaskList';
import Task from '../Task';

const testProps = {
  actions: {},
  tasks: tasksSeed.tasks,
  showCompleted: true
};

describe('TaskList', () => {
  const totalTasksCount = testProps.tasks.length;
  const inCompleteTasksCount = testProps.tasks.filter(t => !t.isComplete)
    .length;

  it('renders correctly', () => {
    const tree = renderer.create(<TaskList {...testProps} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it(`renders ${totalTasksCount} <Task/> components when showCompleted is 'true'`, () => {
    const wrapper = shallow(<TaskList {...testProps} />);
    expect(wrapper.find(Task).length).toEqual(totalTasksCount);
  });

  it(`renders ${inCompleteTasksCount} <Task/> components when showCompleted is 'false'`, () => {
    const props = { ...testProps, showCompleted: false };
    const wrapper = shallow(<TaskList {...props} />);
    expect(wrapper.find(Task).length).toEqual(inCompleteTasksCount);
  });
});
