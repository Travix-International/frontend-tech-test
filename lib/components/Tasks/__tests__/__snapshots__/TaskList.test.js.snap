// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskList renders correctly 1`] = `
<div
  className="root"
>
  <ul
    className="list"
  >
    <li
      className="root "
    >
      <div
        className="checkbox "
      >
        <label
          htmlFor="todo-1"
        >
          Wash the dishes
        </label>
        <input
          checked={false}
          id="todo-1"
          name="checkbox"
          onChange={[Function]}
          type="checkbox"
        />
      </div>
      <p
        className="description"
      >
        Yet another generic task
      </p>
      <div
        className="options"
      >
        <button
          className="edit"
          onClick={[Function]}
        >
          Edit
        </button>
        <button
          className="delete"
          onClick={[Function]}
        >
          Delete
        </button>
      </div>
    </li>
    <li
      className="root "
    >
      <div
        className="checkbox "
      >
        <label
          htmlFor="todo-3"
        >
          Take over the world
        </label>
        <input
          checked={false}
          id="todo-3"
          name="checkbox"
          onChange={[Function]}
          type="checkbox"
        />
      </div>
      <p
        className="description"
      >
        Send an invite to Gru and Megamind
      </p>
      <div
        className="options"
      >
        <button
          className="edit"
          onClick={[Function]}
        >
          Edit
        </button>
        <button
          className="delete"
          onClick={[Function]}
        >
          Delete
        </button>
      </div>
    </li>
    <li
      className="root complete"
    >
      <div
        className="checkbox checked"
      >
        <label
          htmlFor="todo-2"
        >
          Do the laundry
        </label>
        <input
          checked={true}
          id="todo-2"
          name="checkbox"
          onChange={[Function]}
          type="checkbox"
        />
      </div>
      <p
        className="description"
      >
        Existence is pain
      </p>
      <div
        className="options"
      >
        <button
          className="edit"
          onClick={[Function]}
        >
          Edit
        </button>
        <button
          className="delete"
          onClick={[Function]}
        >
          Delete
        </button>
      </div>
    </li>
  </ul>
</div>
`;
